src/server/game/Entities/Object/Object.cpp

```diff
@@ -1422,33 +1422,28 @@ void WorldObject::UpdateAllowedPositionZ(float x, float y, float &z) const
 
     if (Unit const* unit = ToUnit())
     {
-        if (!unit->CanFly())
-        {
-            bool canSwim = unit->CanSwim();
-            float ground_z = z;
-            float max_z;
-            if (canSwim)
-                max_z = GetMapWaterOrGroundLevel(x, y, z, &ground_z);
-            else
-                max_z = ground_z = GetMapHeight(x, y, z);
+        // Unit is flying, ignore.
+        if (unit->IsFlying())
```

Comment:
You need to revisit the logic in this part here. The changes break hover ground z coordinates for units that can fly and have ground hover movement (for example the valkyrs in ICC). So with the changes made here the valkyrs for example never get a proper ground Z returned because they can fly and they can hover above the ground. The IsFlying check is effectively disabling all hover checks in this case so the valkyr can fly into the ground ignoring the hover offset entirely.